### ISSUE TITLE
[Major annoying] Bitwarden does not obey autocomplete="off"

### ISSUE THREAD

**Issue by timint** [bitwarden/browser#1469]
I have an issue that Bitwarden for Chrome does not obey autocomplete=off for input fields. This is very annoying for example when moderating other user accounts and it resets their username to my own.

`<input type="password" autocomplete="Off" />`

**Comment by SergeantConfused** [comment]
Hi timint,

Are you sure it's the Bitwarden extension?
Can you please check if you've disabled Chrome's auto-fill features?
Can you test this in Firefox, please?


Thank you in advance,

**Comment by timint** [comment]
I am sure this is bitwarden

![image](https://user-images.githubusercontent.com/359192/100002003-40dd0a80-2dc4-11eb-8738-73ad0755fa67.png)

**Comment by timint** [comment]
@SergeantConfused I can confirm the problem is in both Firefox and Chrome.
Bitwarden does not obey autocomplete="off".

**Comment by timint** [comment]
In autofill.js

 **Find** 

`        function fillTheElement(el, op) {`

 **After that, add** 

`            if (el.getAttribute('autocomplete') != '' && el.getAttribute('autocomplete').match(/^(off|0|false)$/i)) return false;`

We should also access the `<form>` element if it flagged for no autocompletion.

**Comment by wusatosi** [comment]
I've had a look-around, I think I know what we need to do to fix this, I'll try to implement a fix this week.

**Comment by timint** [comment]
@wusatosi Looking forward to it. Other than that I feel that the prefilling of passwords matched sites better than LastPass does it. Nicely done!

**Comment by wusatosi** [comment]
There's two parts of the problem.
First, the content script does not collect documents with the special attribute correctly.
https://github.com/bitwarden/browser/blob/83cea49a1b9eeb382cf6cb8a7b4b725d0371e3c9/src/content/autofill.js#L300
This collects the value of the auto-complete preference to the background service if its associated attribute is not equal to "off", which is a logic error, we **should** collect this preference when the HTML explicitly disfavors auto complete operation.

Second, the background service is not handling the case of autocomplete=off.

**Comment by wusatosi** [comment]
@timint Can you provide a website that uses autocomplete=off? We can use that as a testing site.

**Comment by timint** [comment]
@wusatosi Here is one that uses `<form autocomplete="off">`:

http://demo.litecart.net/admin/?app=users&doc=edit_user
http://demo.litecart.net/admin/?app=customers&doc=edit_customer

The platform is reset every half hour.

**Comment by wusatosi** [comment]
> @wusatosi Here is one that uses `<form autocomplete="off">`:
> 
> http://demo.litecart.net/admin/?app=users&doc=edit_user
> http://demo.litecart.net/admin/?app=customers&doc=edit_customer
> 
> The platform is reset every half hour.
> 
> I will see if I have a link for `<input autocomplete="off" />` as well.

I didn't see any form with autocomplete="off"
![image](https://user-images.githubusercontent.com/26424577/103462751-6285c600-4d62-11eb-9816-3358b80764c9.png)

**Comment by timint** [comment]
You are supposed to login, lol. I was thinking I should say anything but thought it was obvious. 😁

**Comment by wusatosi** [comment]
oof, I'll test it out later

--------------原始邮件--------------
发件人："T. Almroth "<notifications@github.com&gt;;
发送时间：2021年1月3日(星期天) 凌晨4:10
收件人："bitwarden/browser" <browser@noreply.github.com&gt;;
抄送："wusatosi "<notifications@wusatosi.com&gt;;"Mention "<mention@noreply.github.com&gt;;
主题：Re: [bitwarden/browser] [Major annoying] Bitwarden does not obey autocomplete="off" (#1469)
-----------------------------------

 

 
You are supposed to login, lol. I was thinking I should say anything but thought it was obvious.

**Comment by wusatosi** [comment]
oof, I'll test it out later

**Comment by wusatosi** [comment]
@timint My fix works for the provided URLs on my side. Can you test it from your side?

Unzip this file to a directory and import the directory it from your browser as an extension, under "extension" => "load unpacked"
[bitwarden-autocomplete-fix-v1.zip](https://github.com/bitwarden/browser/files/5765214/bitwarden-autocomplete-fix-v1.zip)

**Comment by wusatosi** [comment]
In the "edit customer page," the email address input is not marked as `autocomplete="off"` btw.
![image](https://user-images.githubusercontent.com/26424577/103548786-43617280-4ee1-11eb-8a07-772ca761b294.png)

**Comment by timint** [comment]
Lol. Yeah, that's what I said. This one uses `<form autocomplete="off">`. Which means all fields inherit this setting, not just one.

**Comment by kspearrin** [comment]
Ignoring autofill for autocomplete="off" fields will break a whole host of websites that users expect to autofill. For example, banks often will put autocomplete="off" on password fields so prevent the browser from saving the users password. Users still expect Bitwarden to autofill these fields.

Here's one example: https://www.wellsfargo.com/

**Comment by cscharf** [comment]
Hi all, I'm going to close this issue. Big thanks to @wusatosi for submitting a PR to change this behavior, however at this time this is expected behavior as Kyle mentioned above and won't necessarily be changed. You can use custom fields for specific sites you want to not auto-fill. I would welcome someone to post in the [Bitwarden Community Forums](https://community.bitwarden.com) for a feature request perhaps for adding a configuration setting/toggle or something to that effect (that doesn't change the current default and expected behavior). Github however is for bugs/issues and this would require more thought around an enhancement to the product.

**Comment by timint** [comment]
@kspearrin I don't care what your bank does. But if it uses regular passwords to protect your money, you probably should worry about other things than having your most precious password not being prefilled.

Mainting CRM systems or administrating user profiles is not something out of the ordinary. Bitwarden becomes totally useless for this. Everytime I open up a customer profile, their email and password is replaced and I have to retype their email and clear out the new password. How can you say this is the expected behavior. The other password managers e.g. LastPass don't have this behavior. Not even your browser does.

What are you thinking, really?

**Comment by PabloRomanH** [comment]
This is still broken. I'm trying to change my email address in a website (freepik.com), and as soon as I enter into my user profile, Bitwarden automatically fills in 5 different fields (username, email, new password, current password, current password again), when the only two that would be reasonably expected to autofill would be current password and current password again. I can't even see my current username. We need to at least be given the option of disabling autofill for certain URLs.

**Comment by timint** [comment]
@PabloRomanH I tried telling them. They won't listen to logic. Websites implement `autocomplete="off"` for a reason. But they are sacrificing that just so they can have their bank login prefilled 😮.

They should turn it around the other way, make a list of urls that don't obey autocomplete=off. So anyone crazy enough to prefill their bank login could continue doing so. But save the rest of the planet. 😁

**Comment by phuze** [comment]
LastPass offers application owners the option, by suppling a data attribute, `lpignore`, for example:

```
<input type="password" name="password" autocomplete="off" data-lpignore="true" />
```

It'd be nice if BitWarden supported something similar, for example:

```
<input type="password" name="password" autocomplete="off" data-bwignore="true" />
```

**Comment by genechuah** [comment]
Bump. Still broken. Should not behave this way. Bad add-on, bad.

**Comment by timint** [comment]
> For example, banks often will put autocomplete="off" on password fields so prevent the browser from saving the users password. Users still expect Bitwarden to autofill these fields.

Yes, and they backed it up with this sentence:

> For example, banks often will put autocomplete="off" on password fields so prevent the browser from saving the users password. Users still expect Bitwarden to autofill these fields.

Hmm.. let's prefill the password for banks.. what can go wrong.. hmm... 😆
This is not a feature. This is one major reason to stop using Bitwarden.

**Comment by phuze** [comment]
As the user who's chosen to use BitWarden, and opted-in to the "auto-fill" feature, it's your responsibility to manage your settings. If the feature is causing you trouble, you have the option of turning it off.

Alternatively, use the available options to disable the auto-fill feature on the specific page where you don't want this to happen -- something along the lines of:

![image](https://user-images.githubusercontent.com/9218035/206587516-ce26c151-e285-4fc1-ae9d-33f50b2de0c5.png)

It is short-sighted to suggest that BitWarden is broken, because it doesn't behave in a manner that would allow you to enforce your opinion on others. Many browsers, Chrome and Firefox included, [do not support](https://developer.mozilla.org/en-US/docs/Web/Security/Securing_your_site/Turning_off_form_autocompletion#the_autocomplete_attribute_and_login_fields:~:text=For%20this%20reason%2C%20many%20modern%20browsers%20do%20not%20support%20autocomplete%3D%22off%22%20for%20login%20fields%3A) `autocomplete="off"` as they prioritize users over developers -- as it should be, because **user choice** matters more.

@kspearrin @djsmith85 consider locking this thread.

**Comment by genechuah** [comment]
I'm writing this as a developer and a user has reported the issue but we didn't know it was BitWarden until some troubleshooting. Other users may not be so lucky and will encounter puzzling behavior on sites where the developer has specified autocomplete to be turned off.

**Comment by genechuah** [comment]
Not all users would be tech-savvy enough to figure out that it's Bitwarden causing the issue.

**Comment by phuze** [comment]
> I'm writing this as a developer and a user has reported the issue but we didn't know it was BitWarden until some troubleshooting. Other users may not be so lucky and will encounter puzzling behavior on sites ...

[PEBCAK](https://en.wikipedia.org/wiki/User_error). This is not, nor should be, a concern of the BitWarden application. It's also not unique to BitWarden. Other password managers, such as Dashlane and LastPass, also ignore the autocomplete attribute -- though Lastpass does offer an option to respect the autocomplete attribute if the user chooses to enable that.

> ... where the developer has specified autocomplete to be turned off.

User choice matters more.

> Not all users would be tech-savvy enough to figure out that it's Bitwarden causing the issue.

If you find that this is creating issue for you, or your support teams, consider including a notice on affected pages, informing users of the possibility that this could occur when using password managers. Providing a help document which guides your users on how to configure their password manager to prevent autofilling on certain pages, would be added value.

Where appropriate, you may also consider the [new-password](https://developer.mozilla.org/en-US/docs/Web/Security/Securing_your_site/Turning_off_form_autocompletion#preventing_autofilling_with_autocompletenew-password) value for the autocomplete attribute.

As a developer, respect the choice your users have made.

**Comment by timint** [comment]
So we should cross a red light because everyone else is crossing the red light?

I believe autocomplete (part of HTML 5.2) was added to control the behavior of form fillers because enough users demanded to solve a problem. I believe that webmasters add the attribute autocomplete="off" for reasons, not ignorance. It could be out of security aspects or there could be use cases when there were problems with form fillers. Yes, I really believe that webmasters kind of has it thought through when they use it, also with the end-users in mind. What other motivations could there be for adding it?

Looking back at my own projects. I have added autocomplete="off" because there were reports of a problem with form fillers.

Just rebelliously going against autocomplete="off" by filling input elements anyway is a somehow aggressive strategy. A way to satisy both the world of trespassers, and the world of obeyers could be a hybrid solution. How about not prefilling the field and honour autocomplate="off", but leave a form helper button when focusing on the input field. that lets the user choose themselves to override by prefilling?

The use intentions for the implementation of autocomplete is available here:
https://www.w3.org/WAI/WCAG21/Techniques/html/H98

**Comment by timint** [comment]
@phuze Thank you for pointing out `autocomplete="new-password"`. It might have a better outcome than `off` with some fillers.

**Comment by HankAviator** [comment]
It might be a good idea to add an option to allow users who ...
> So we should cross a red light because everyone else is crossing the red light?

... will obey `autocomplete` field.
However, websites in my country tend to forbid users from using password managers and auto password completion, as a lazy  effort to 'be safe', which is totally not the case.

**Comment by notken** [comment]
Actually autocomplete="new-password" doesn't work either and is being ignored by Bitwarden.

I've had reports that users of our site have had their email address leaked into social network fields. When I test it using a profile update page, I have two new-password fields (change  it and confirm it), and they are both being pre-filled with my password. A field called "Website" with autocomplete="off" is being pre-filled with the username/email address, which users are *not noticing*.

It's not only wiping out data, it's exposing personal information in fields that users have said are fine to be public.

I agree with other developers here that blindly ignoring autocomplete values and expecting users to turn that behaviour off is wrong and dangerous. Many users are not savvy or observant enough to spot this is happening. I've just searched my site's database and found over 100 users with email addresses exposed in social network fields.

**Comment by Itzdlg** [comment]
Seriously? If you won't accept autocomplete="off", then accept an autocomplete="new-password" and "new-email"?? Doesn't make sense that this is the "target behavior" since it's auto-filling the administrator's information on create modals. 
(My own information, mind you, as I'm a developer using this extension which has caused me 2 separate problems in development in the past week alone. Seems only "non-saavy" users are the target demographic.)

**Comment by flixx** [comment]
@kspearrin @cscharf @djsmith85 

Is there any other way for a developer to make sure that some fields in some admin-create-form don't get autofilled with the administrators login password? We see some less protected fields in our django admin interface being auto-filled with admin-passwords. Any way to prevent this?

**Comment by alcohol** [comment]
I can understand the reasoning behind ignoring `autocomplete="off"`, but as a user I sometimes find it annoying/frustrating that it does not respect the distinction between `autocomplete="new-password"` and `autocomplete="current-password"` when I'm trying to change my password and I have to provide both a new password and my current password.

**Comment by frostyfrog** [comment]
> > ... where the developer has specified autocomplete to be turned off.
> 
> User choice matters more.

In my experience, this can actively harm users by, for instance, populating one's password into a search field that automatically sends the search criteria to the server, making it into the logs for things such as analytics. And developers have NO WAY to prevent this security problem.

Case in point, I'm working on a form that has a search bar with it's own autocomplete/autofill (like Google's search bar), and BitWarden keeps choosing to enter information into said field, and there's nothing I can do to prevent it. Meaning, if I choose to have analytics to understand what people are searching for, a user's "secure" information would constantly make it into the logs & analytics.

**Comment by timint** [comment]
One of the claims that it's a good idea to continue disobeying autocomplete="off" is that users expect their bank login to be prefilled.

As stated by [kspearrin](https://github.com/kspearrin), I quote:

> Ignoring autofill for autocomplete="off" fields will break a whole host of websites that users expect to autofill. For example, banks often will put autocomplete="off" on password fields so prevent the browser from saving the users password. Users still expect Bitwarden to autofill these fields.
> 
> Here's one example: https://www.wellsfargo.com/

Now.. what could go wrong with that?

![TheodorosGIF](https://github.com/bitwarden/clients/assets/359192/199b3b36-ff9f-4421-a28e-a57451366347)

**Comment by tomastan** [comment]
We face a similiar problem with user management webapp (call it CRM).

On a profile edit page there are fields like email, name, and a password reset. There were numerous times, when editing just a name of the user, overrides his password and email with admin's auto-filled password, if the admin uses Bitwarden.  Ofc, autofill could be disabled, but in this case admin's own login/pass would not autofilled on his own login.

If respecting autocomplete="off" is too disputable step for Bitwarden team, why not to make a compromise and add an option to "Respect autocomplete="off"" for users who care? (Like LastPass does, if I understand from other comments).

**Comment by Gravestam** [comment]
This is indeed really annoying. It should, imo, to be an option i bitwarden.  For the best user experience and to accommodate for the kspearrin quote above it should be implemented as an option per bitwarden login item. imo

**Comment by clintebbesen** [comment]
for me Bitwarden is trying to auto complete on all fields on a form and I don't want it on any, including drop-down picking dates, how can I prevent Bitwarden from adding to every field in the form and prevent it entirely?

![image](https://github.com/bitwarden/clients/assets/150989350/a67b74b1-9347-4397-9c48-acf9873ed08b)
![image](https://github.com/bitwarden/clients/assets/150989350/4802e6bc-ac19-44df-b095-fe825766c504)

**Comment by mrforsythexeter** [comment]
Can it just be made an option please in the browser extension, you are annoying the people (developers) that may recommend the product to your user which will pay for it. I tend to operate "I can't use myself as an example". So ok, continue to ignore that setting (wrong).. however give the people that know what they are doing, the option to choose to have obey the attribute.

**Comment by TargiX** [comment]
Yeah, in our company web site we have similar issue, we have field for password that you can change to new if you want, but it constantly prefills it and we getting falsy attempts to change that password, you guys are really crazy!

**Comment by timint** [comment]
I vote reopen!

>> For example, banks often will put autocomplete="off" on password fields so prevent the browser from saving the users password. Users still expect Bitwarden to autofill these fields.
>> 
> Hmm.. let's prefill the password for banks.. what can go wrong.. hmm... 😆

Someone remind management why there are passwords. Again.. this is a disaster.

**Comment by italiangamesfactory** [comment]
This is still happening, even in the newly refreshed Chrome extension.

**Comment by fulvio-m** [comment]
This is still happening, even in the newly refreshed Chrome extension.

**Comment by vktrl** [comment]
Also facing this issue. autocomplete="off" should mean exactly that. It shouldn't be an option. It's driving me nuts right now because I have inputs in my application that Bitwarden wants to autocomplete but should never be autocompleted. Not for me, not for anyone else.

**Comment by mrforsythexeter** [comment]
> > Good news everyone! The latest version of the browser extension supports [blocking autofill](https://bitwarden.com/help/blocking-uris/) by domain! We don't have any plans right now to add support for "off" as a value for the autocomplete property. But, blocked domains should cover most of the use cases mentioned in this thread.
> 
> Unfortunately it does not. The user could do that before (block URL or set masks), nothing new here. If you block a domain, you cannot use auto-fill to the system at all. The problem though is not with a whole website, but with _specific_ fields on the site Bitwarden is eager to fill.
> 
> _Hence, this is not relevant for us anymore, as we ultimately switched to ProtonPass tired of this issue._

I completely agree with this 100%. Honour the web developers request, if a user choses to manually use the input to complete a field, that is up to them, not the developer.

**Comment by timint** [comment]
I agree with everyone else here. "blocking autofill by domain" really does nothing to improve this situation. There are several reasons why a specific form should not be prefilled. And the best reason might be the instruction the developer implemented. Forms are autocompletable by default. Why  would developers separately flag forms to block autocomplete?

Are we still arguing about this just because management of this repository use this extension to have their bank credentials prefilled?  Using somone else's computer and all bank credentials are filled in, isn't that kind of begging for a legal case in court?

You could at least back this feature up with some reasonable logic. We provided our logic for why we are against it.

**Comment by vmyazin** [comment]
Please address ASAP! Add `bitwarden-ignore` or something.

**Comment by danielnitu** [comment]
We just stumbled into this issue as well. `autocomplete="off"` is just ignored by Bitwarden. It's amazing how this has been opened for 5 years and the devs still can't be convinced to respect simple HTML protocols.

**Comment by tiagorangel1** [comment]
switched to proton pass because of this

**Comment by LosD** [comment]
> Good news everyone! The latest version of the browser extension supports [blocking autofill](https://bitwarden.com/help/blocking-uris/) by domain! We don't have any plans right now to add support for "off" as a value for the autocomplete property. But, blocked domains should cover most of the use cases mentioned in this thread.
> 
> [@clintebbesen](https://github.com/clintebbesen) your issue with forms autocomplete sounds like a different bug. If you're still seeing this behavior: could you create a separate issue with more information about your situation?

This is so damn stupid. Why would I block a domain, when the issue is on specific forms? I still want to have login support, but user adminstration, or anything involving users or emails is hell with this.

Yes, you have "data-bwignore", but extension-specific attributes is terrible. Currently, our user selection (not related to login, but triggers Bitwarden, because "email" is in part of the field description 🙄 ) has these attributes:
`autocomplete="off" data-1p-ignore data-lpignore="true" data-form-type="other" data-bwignore`, because all password managers thinks that they know what a field is for. So. Damn. Stupid.

Anyone know a PW manager that isn't annoying by design?

**Comment by tomastan** [comment]
> Anyone know a PW manager that isn't annoying by design?

Proton pass. Switched to it because of this issue couple of years ago. The downsides it has - limited offline support, and throws ads time to time.

**Comment by LosD** [comment]
> > Anyone know a PW manager that isn't annoying by design?
> 
> Proton pass. Switched to it because of this issue couple of years ago. The downsides it has - limited offline support, and throws ads time to time.

Sorry, no. "Ads" is very much included in "annoying by design". And it seems you can only get the business version with at least 3 users.

**Comment by joezappie** [comment]
This has been a poor stance by browsers and password managers for a decade.

> [Don’t try to fool the browser](https://www.chromium.org/developers/design-documents/create-amazing-password-forms/#dont-try-to-fool-the-browser)
> Password managers (either built into the browser, or external) are designed to ease the user experience. Inserting fake fields, using incorrect autocomplete attributes or taking advantage of the weaknesses of the existing password managers simply leads to frustrated users.

You know whats frustrating as a user? When I have a field that is not a password/email or address but yet my password manager still pops up and autofills unrelated fields and I have to go back and delete them, instead of just letting a developer notate what field data that should be? We only have 1 field type that lets the data be obscured - `password` and so its used for 100 other things other than passwords so respect developers intentions with their forms as they know better than a password manager on what data should go where. 

tldr: My hidden API key field is not a password.

**Comment by LosD** [comment]
> This has been a poor stance by browsers and password managers for a decade.

It had its place a while ago, when terrible website developers had a silly idea about their sites being more secure without passwords being stored in pw managers.

But these days, aggressive pw managers is a much bigger problem than silly websites.

**Comment by merlinz01** [comment]
I agree it is essential for Bitwarden to respect `autocomplete="off"`. ~One hack to get around it when it wrongly detects a keyword is to insert a zero-width space in the middle of the keyword, like this: `<input type="text" placeholder="Enter your e&#8203;mail here">`. The invisible character breaks up the word so Bitwarden doesn't detect it but it is visually the same.~ Edit: See my comment below.

**Comment by merlinz01** [comment]
https://github.com/bitwarden/clients/blob/ba1c74b9052a79806e1f159547f4e5cc442227e7/apps/browser/src/autofill/services/collect-autofill-content.service.ts#L70-L74

Judging by these lines and my experimentation, adding `data-bwignore="1"` to the input is what it takes to disable the popup.

**Comment by spryzone** [comment]
I'm compiling all password manager ignore attributes into this gist if it helps anyone. Thanks @merlinz01 for the Bitwarden example.

https://gist.github.com/spryzone/efe664e0564b3c66f2fb09e7207a8d54

**Comment by brc-dd** [comment]
Bit annoyed that there is no consensus between package managers about what that data attribute should be called. The popup recently started opening on a hidden (visually, not type=hidden) input field in Discord web for me (and looks like it happens on other fields too - #18380). They do add attributes for 1password and LastPass, but not Bitwarden. Created a userscript to add `data-bwignore` if any of the other password manager attributes are detected - https://gist.github.com/brc-dd/7ec651ccfe940590bea59a72d2d0d8a0. The `SELECTOR` variable can be modified to include `[autocomplete=off]` in the list if one wants.

**Comment by mattMillerC** [comment]
This is still very annoying, please for the love of God reopen and adopt the standard

**Comment by electricsteve** [comment]
> You know whats frustrating as a user? When I have a field that is not a password/email or address but yet my password manager still pops up and autofills unrelated fields and I have to go back and delete them, instead of just letting a developer notate what field data that should be? We only have 1 field type that lets the data be obscured - `password` and so its used for 100 other things other than passwords so respect developers intentions with their forms as they know better than a password manager on what data should go where.
> 
> tldr: My hidden API key field is not a password.

I am having this exact issue. I want to hide an api key that is entered. That is NOT a password. Also the app will be self-hostable with every instance only having ONE api key so it wouldn't even be useful between user accounts.

**Comment by thibaudcolas** [comment]
Just reported #20320 as another example of this in practice. I found out about `data-bwignore` via [release notes from 2021](https://bitwarden.com/help/releasenotes/#2021-06-29) – does anyone know if there is any official Bitwarden docs where it’s documented? Release notes mention doesn’t inspire confidence that this is here to stay.
